File: projects/java/cdm/src/mil/tatrc/physiology/datamodel/system/equipment/anesthesia/actions/SEAnesthesiaMachineConfiguration.java
/**************************************************************************************
Copyright 2015 Applied Research Associates, Inc.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the License
at:
http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
**************************************************************************************/

package mil.tatrc.physiology.datamodel.system.equipment.anesthesia.actions;

import mil.tatrc.physiology.datamodel.CDMSerializer;
import mil.tatrc.physiology.datamodel.bind.AnesthesiaMachineConfigurationData;
import mil.tatrc.physiology.datamodel.substance.SESubstanceManager;
import mil.tatrc.physiology.datamodel.system.equipment.anesthesia.SEAnesthesiaMachine;

public class SEAnesthesiaMachineConfiguration extends SEAnesthesiaMachineAction
{
	protected SEAnesthesiaMachine configuration;
	protected String              configurationFile;
  
	public SEAnesthesiaMachineConfiguration()
	{
		
	}
	
	public SEAnesthesiaMachineConfiguration(SEAnesthesiaMachineConfiguration other)
	{
		this();
		copy(other);		
	}
	
	public void copy(SEAnesthesiaMachineConfiguration other)
	{
	  if(this==other)
      return;
    super.copy(other);
		this.configuration.copy(other.configuration);
		this.configurationFile=other.configurationFile;
	}
	
	public void reset()
	{
		super.reset();
		
		if (this.configuration != null)
			this.configuration.reset();
		if (this.configurationFile != null)
			this.configurationFile="";
	}
	
	public boolean isValid()
	{
		return hasConfiguration() || hasConfigurationFile();
	}
	
	public boolean load(AnesthesiaMachineConfigurationData in, SESubstanceManager subMgr)
	{
		super.load(in);
		if(in.getConfiguration()!=null)
			getConfiguration().load(in.getConfiguration(),subMgr);
		else if(in.getConfigurationFile()!=null)
		  this.configurationFile=in.getConfigurationFile();
		return isValid();
	}
	
	public AnesthesiaMachineConfigurationData unload()
	{
	  AnesthesiaMachineConfigurationData data = CDMSerializer.objFactory.createAnesthesiaMachineConfigurationData();
		unload(data);
		return data;
	}
	
	protected void unload(AnesthesiaMachineConfigurationData data)
	{
		super.unload(data);
		if(this.hasConfiguration())
		  data.setConfiguration(this.configuration.unload());
		else if(this.hasConfigurationFile())
		  data.setConfigurationFile(this.configurationFile);
	}
	
	public boolean hasConfiguration()
	{
	  return this.configuration!=null;
	}
	public SEAnesthesiaMachine getConfiguration()
	{
	  if(this.configuration==null)
	    this.configuration=new SEAnesthesiaMachine();
		return this.configuration;
	}
	
	public boolean hasConfigurationFile()
	{
	  return this.configurationFile!=null&&!this.configurationFile.isEmpty();
	}
	public String getConfigurationFile()
	{
	  return this.configurationFile;
	}
	public void setConfigurationFile(String s)
	{
	  this.configurationFile = s;
	}
	
	public String toString()
	{
	  String str = "Anesthesia Machine Configuration";
	  if(hasConfiguration())
	  {
	    str += "\n\tConnection: "; str += this.configuration.hasConnection()?this.configuration.getConnection():"Not Supplied";
	    str += "\n\tInlet Flow: "; str += this.configuration.hasInletFlow()?this.configuration.getInletFlow():"Not Supplied";
	    str += "\n\tInspiratory/Expiratory Ratio: "; str += this.configuration.hasInspiratoryExpiratoryRatio()?this.configuration.getInspiratoryExpiratoryRatio():"Not Supplied";
	    str += "\n\tOxygen Fraction: "; str += this.configuration.hasOxygenFraction()?this.configuration.getOxygenFraction():"Not Supplied";    
	    str += "\n\tOxygen Source: "; str += this.configuration.hasOxygenSource()?this.configuration.getOxygenSource():"Not Supplied";
	    str += "\n\tPositive End Expired Pressure: "; str += this.configuration.hasPositiveEndExpiredPressure()?this.configuration.getPositiveEndExpiredPressure():"Not Supplied";
	    str += "\n\tPrimary Gas: "; str += this.configuration.hasPrimaryGas()?this.configuration.getPrimaryGas():"Not Supplied";
	    str += "\n\tReliefValvePressure: "; str += this.configuration.hasReliefValvePressure()?this.configuration.getReliefValvePressure():"Not Supplied";
	    str += "\n\tRespiratory Rate: "; str += this.configuration.hasRespiratoryRate()?this.configuration.getRespiratoryRate():"Not Supplied";
	    str += "\n\tVentilator Pressure: "; str += this.configuration.hasVentilatorPressure()?this.configuration.getVentilatorPressure():"Not Supplied";
	    if(this.configuration.hasLeftChamber())
	    {
	      str += "\n\tLeftChamberState: "; str += this.configuration.getLeftChamber().hasState()?this.configuration.getLeftChamber().getState():"Not Supplied";
	      str += "\n\tLeftChamberSubstance: "; str += this.configuration.getLeftChamber().hasSubstance()?this.configuration.getLeftChamber().getSubstance().getName():"Not Supplied";    
	      str += "\n\tLeftChamberSubstanceAmount: "; str += this.configuration.getLeftChamber().hasSubstanceFraction()?this.configuration.getLeftChamber().getSubstanceFraction():"Not Supplied";	         
	    }
	    if(this.configuration.hasRightChamber())
      {
        str += "\n\tRightChamberState: "; str += this.configuration.getRightChamber().hasState()?this.configuration.getRightChamber().getState():"Not Supplied";
        str += "\n\tRightChamberSubstance: "; str += this.configuration.getRightChamber().hasSubstance()?this.configuration.getRightChamber().getSubstance().getName():"Not Supplied";    
        str += "\n\tRightChamberSubstanceAmount: "; str += this.configuration.getRightChamber().hasSubstanceFraction()?this.configuration.getRightChamber().getSubstanceFraction():"Not Supplied";   
      }
	    if(this.configuration.hasOxygenBottleOne())
	    {
	      str += "\n\tOxygenBottleOneVolume: "; str += this.configuration.getOxygenBottleOne().hasVolume()?this.configuration.getOxygenBottleOne().getVolume():"Not Supplied";
	    }
	    if(this.configuration.hasOxygenBottleTwo())
      {
        str += "\n\tOxygenBottleTwoVolume: "; str += this.configuration.getOxygenBottleTwo().hasVolume()?this.configuration.getOxygenBottleTwo().getVolume():"Not Supplied";
      }
	  }
		if(this.hasConfigurationFile())
		  str +="\n\tFile: "+this.configurationFile;
		return str;
	}
}

File: projects/biogears/src/cdm/system/equipment/Anesthesia/actions/SEAnesthesiaMachineConfiguration.cpp
/**************************************************************************************
Copyright 2015 Applied Research Associates, Inc.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the License
at:
http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
**************************************************************************************/
#include <biogears/cdm/system/equipment/Anesthesia/actions/SEAnesthesiaMachineConfiguration.h>

#include <biogears/cdm/properties/SEScalarFraction.h>
#include <biogears/cdm/properties/SEScalarFrequency.h>
#include <biogears/cdm/properties/SEScalarPressure.h>
#include <biogears/cdm/properties/SEScalarVolume.h>
#include <biogears/cdm/properties/SEScalarVolumePerTime.h>
#include <biogears/cdm/substance/SESubstance.h>
#include <biogears/cdm/substance/SESubstanceManager.h>
#include <biogears/cdm/system/equipment/Anesthesia/SEAnesthesiaMachine.h>
#include <biogears/cdm/system/equipment/Anesthesia/SEAnesthesiaMachineChamber.h>
#include <biogears/cdm/system/equipment/Anesthesia/SEAnesthesiaMachineOxygenBottle.h>

namespace biogears {
SEAnesthesiaMachineConfiguration::SEAnesthesiaMachineConfiguration(SESubstanceManager& substances)
  : SEAnesthesiaMachineAction()
  , m_Substances(substances)
{
  m_Configuration = nullptr;
  InvalidateConfigurationFile();
}
//-----------------------------------------------------------------------------
SEAnesthesiaMachineConfiguration::~SEAnesthesiaMachineConfiguration()
{
  Clear();
}
//-----------------------------------------------------------------------------
void SEAnesthesiaMachineConfiguration::Clear()
{
  SEAnesthesiaMachineAction::Clear();
  InvalidateConfigurationFile();
  SAFE_DELETE(m_Configuration);
}
//-----------------------------------------------------------------------------
bool SEAnesthesiaMachineConfiguration::IsValid() const
{
  return SEAnesthesiaMachineAction::IsValid() && (HasConfiguration() || HasConfigurationFile());
}
//-----------------------------------------------------------------------------
bool SEAnesthesiaMachineConfiguration::Load(const CDM::AnesthesiaMachineConfigurationData& in)
{
  SEAnesthesiaMachineAction::Load(in);
  if (in.ConfigurationFile().present())
    SetConfigurationFile(in.ConfigurationFile().get());
  if (in.Configuration().present())
    GetConfiguration().Load(in.Configuration().get());
  return true;
}
//-----------------------------------------------------------------------------
CDM::AnesthesiaMachineConfigurationData* SEAnesthesiaMachineConfiguration::Unload() const
{
  CDM::AnesthesiaMachineConfigurationData* data = new CDM::AnesthesiaMachineConfigurationData();
  Unload(*data);
  return data;
}
//-----------------------------------------------------------------------------
void SEAnesthesiaMachineConfiguration::Unload(CDM::AnesthesiaMachineConfigurationData& data) const
{
  SEAnesthesiaMachineAction::Unload(data);
  if (HasConfiguration())
    data.Configuration(std::unique_ptr<CDM::AnesthesiaMachineData>(m_Configuration->Unload()));
  else if (HasConfigurationFile())
    data.ConfigurationFile(m_ConfigurationFile);
}
//-----------------------------------------------------------------------------
bool SEAnesthesiaMachineConfiguration::HasConfiguration() const
{
  return m_Configuration != nullptr;
}
//-----------------------------------------------------------------------------
SEAnesthesiaMachine& SEAnesthesiaMachineConfiguration::GetConfiguration()
{
  m_ConfigurationFile = "";
  if (m_Configuration == nullptr)
    m_Configuration = new SEAnesthesiaMachine(m_Substances);
  return *m_Configuration;
}
//-----------------------------------------------------------------------------
const SEAnesthesiaMachine* SEAnesthesiaMachineConfiguration::GetConfiguration() const
{
  return m_Configuration;
}
//-----------------------------------------------------------------------------
const char* SEAnesthesiaMachineConfiguration::GetConfigurationFile() const
{
  return m_ConfigurationFile.c_str();
}
//-----------------------------------------------------------------------------
void SEAnesthesiaMachineConfiguration::SetConfigurationFile(const char* fileName)
{
  SetConfigurationFile(std::string{ fileName });
}
//-----------------------------------------------------------------------------
void SEAnesthesiaMachineConfiguration::SetConfigurationFile(const std::string& fileName)
{
  if (m_Configuration != nullptr)
    SAFE_DELETE(m_Configuration);
  m_ConfigurationFile = fileName;
}
//-----------------------------------------------------------------------------
bool SEAnesthesiaMachineConfiguration::HasConfigurationFile() const
{
  return m_ConfigurationFile.empty() ? false : true;
}
//-----------------------------------------------------------------------------
void SEAnesthesiaMachineConfiguration::InvalidateConfigurationFile()
{
  m_ConfigurationFile = "";
}
//-----------------------------------------------------------------------------
void SEAnesthesiaMachineConfiguration::ToString(std::ostream& str) const
{
  str << "Anesthesia Machine Configuration";
  if (HasComment())
    str << "\n\tComment: " << m_Comment;
  if (HasConfigurationFile())
    str << "\n\tConfiguration File: ";
  str << m_ConfigurationFile;
  if (HasConfiguration()) {
    str << "\n\tConnection: ";
    m_Configuration->HasConnection() ? str << m_Configuration->GetConnection() : str << "Not Set";
    str << "\n\tInlet Flow Rate: ";
    m_Configuration->HasInletFlow() ? str << m_Configuration->GetInletFlow() : str << "NaN";
    str << "\n\tInspiratoryExpiratory Ratio: ";
    m_Configuration->HasInspiratoryExpiratoryRatio() ? str << m_Configuration->GetInspiratoryExpiratoryRatio() : str << "NaN";
    str << "\n\tOxygen Fraction: ";
    m_Configuration->HasOxygenFraction() ? str << m_Configuration->GetOxygenFraction() : str << "NaN";
    str << "\n\tOxygen Source: ";
    m_Configuration->HasOxygenSource() ? str << m_Configuration->GetOxygenSource() : str << "Not Set";
    str << "\n\tPositive End Expired Pressure: ";
    m_Configuration->HasPositiveEndExpiredPressure() ? str << m_Configuration->GetPositiveEndExpiredPressure() : str << "NaN";
    str << "\n\tPrimary Gas: ";
    m_Configuration->HasPrimaryGas() ? str << m_Configuration->GetPrimaryGas() : str << "Not Set";
    str << "\n\tRelief Valve Pressure: ";
    m_Configuration->HasReliefValvePressure() ? str << m_Configuration->GetReliefValvePressure() : str << "NaN";
    str << "\n\tRespiratory Rate: ";
    m_Configuration->HasRespiratoryRate() ? str << m_Configuration->GetRespiratoryRate() : str << "NaN";
    str << "\n\tVentilator Pressure: ";
    m_Configuration->HasVentilatorPressure() ? str << m_Configuration->GetVentilatorPressure() : str << "NaN";
    if (m_Configuration->HasLeftChamber()) {
      str << "\n\tLeftChamberState: ";
      m_Configuration->GetLeftChamber().HasState() ? str << m_Configuration->GetLeftChamber().GetState() : str << "Not Supplied";
      str << "\n\tLeftChamberSubstance: ";
      m_Configuration->GetLeftChamber().HasSubstance() ? str << m_Configuration->GetLeftChamber().GetSubstance()->GetName() : str << "Not Set";
      str << "\n\tLeftChamberSubstanceAmount: ";
      m_Configuration->GetLeftChamber().HasSubstanceFraction() ? str << m_Configuration->GetLeftChamber().GetSubstanceFraction() : str << "Not Set";
    }
    if (m_Configuration->HasRightChamber()) {
      str << "\n\tRightChamberState: ";
      m_Configuration->GetRightChamber().HasState() ? str << m_Configuration->GetRightChamber().GetState() : str << "Not Supplied";
      str << "\n\tRightChamberSubstance: ";
      m_Configuration->GetRightChamber().HasSubstance() ? str << m_Configuration->GetRightChamber().GetSubstance()->GetName() : str << "Not Set";
      str << "\n\tRightChamberSubstanceAmount: ";
      m_Configuration->GetRightChamber().HasSubstanceFraction() ? str << m_Configuration->GetRightChamber().GetSubstanceFraction() : str << "Not Set";
    }
    if (m_Configuration->HasOxygenBottleOne()) {
      str << "\n\tOxygenBottleOneVolume: ";
      m_Configuration->GetOxygenBottleOne().HasVolume() ? str << m_Configuration->GetOxygenBottleOne().GetVolume() : str << "Not Set";
    }
    if (m_Configuration->HasOxygenBottleTwo()) {
      str << "\n\tOxygenBottleTwoVolume: ";
      m_Configuration->GetOxygenBottleTwo().HasVolume() ? str << m_Configuration->GetOxygenBottleTwo().GetVolume() : str << "Not Set";
    }
  }
  str << std::flush;
}
  //-----------------------------------------------------------------------------
}